src/librustc_lint/types.rs

```diff
@@ -531,23 +531,23 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         match ty.kind {
             ty::FnPtr(_) => true,
             ty::Ref(..) => true,
-            ty::Adt(field_def, substs) if field_def.repr.transparent() && !field_def.is_union() => {
-                for field in field_def.all_fields() {
-                    let field_ty = self.cx.tcx.normalize_erasing_regions(
-                        self.cx.param_env,
-                        field.ty(self.cx.tcx, substs),
-                    );
-                    if field_ty.is_zst(self.cx.tcx, field.did) {
```

Comment:
What does `is_zst` even do? I'm surprised to see that method on `Ty`, and suspicious of how it might be implemented.

Reply:
I added it in an earlier PR (or I just made it with existing logic, I forget), it’s predominantly used in lints which care about transparent types. Its logic is the same as where transparent types are checked in typeck (but it wasn’t trivial to factor it out there too) - it basically gets the layout of the type and calls `is_zst` on that, returning `false` if any of the intermediate stuff fails.

Reply:
I don't think `Ty` should have methods that interact with layout, feels like an abstraction violation.
If something relies on layout that should be clear (i.e. `is_zst` callers should use `layout_of` themselves - not to mention that  e.g. `LateLintContext` should have better `layout_of` integration than doing it on `Ty`)